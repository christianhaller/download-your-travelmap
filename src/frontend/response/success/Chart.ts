import type { EnhancedPin } from "../../../backend/interace";
import load from "load-js/src/load-js.js";

const getOptions = (width, pointSize): google.visualization.GeoChartOptions => {
  return {
    backgroundColor: {
      fill: "transparent",
      stroke: "#FFF",
      strokeWidth: 0,
    },
    datalessRegionColor: "#f5f5f5",
    pointSize,
    displayMode: "markers",
    resolution: "countries",
    sizeAxis: { minValue: 1, maxValue: 200 },
    legend: "none",
    keepAspectRatio: true,
    width,
    colorAxis: { minValue: 1, maxValue: 10, colors: ["#ff0099", "#000"] },
  };
};

const getData = (places: EnhancedPin[]) => {
  const data = new google.visualization.DataTable();
  data.addColumn("number", "Lat");
  data.addColumn("number", "Long");
  data.addColumn("number", "Value");
  data.addColumn({ type: "string", role: "tooltip" });

  places.forEach(({ lat, lng, city, flags }) => {
    data.addRows([[lat, lng, flags.includes("been") ? 1 : 10, city]]);
  });
  return data;
};

export class Chart {
  private places: EnhancedPin[];
  private el: HTMLElement;
  private chart: google.visualization.GeoChart;

  private load() {
    return new Promise(async (resolve) => {
      await load({
        url: "https://www.gstatic.com/charts/loader.js",
        async: true,
      });

      google.charts.load("current", {
        packages: ["geochart"],
      });
      google.charts.setOnLoadCallback(resolve);
    });
  }

  async init(parent: HTMLElement, places: EnhancedPin[]) {
    this.el = parent.querySelector("#chart");
    this.places = places;
    await this.load();

    this.chart = new google.visualization.GeoChart(this.el);

    this.chart.draw(getData(this.places), getOptions(this.el.offsetWidth - 40));
    return this;
  }
  async getImage(): Promise<Blob> {
    return new Promise((resolve) => {
      const chart = new google.visualization.GeoChart(
        document.getElementById("image")
      );

      google.visualization.events.addListener(chart, "ready", async () => {
        const res = await fetch(chart.getImageURI());
        resolve(res.blob());
      });
      chart.draw(getData(this.places), getOptions(3000, 100));
    });
  }
}
