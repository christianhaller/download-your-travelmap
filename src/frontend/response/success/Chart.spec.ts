import { Chart } from "./Chart";

jest.mock("load-js/src/load-js.js", () => jest.fn().mockResolvedValue(true));
let draw;

describe("Chart", () => {
  window.google = {
    visualization: {
      // @ts-ignore
      events: {
        // @ts-ignore
        addListener: (el, name, cb) => {
          cb();
        },
      },
      GeoChart: jest.fn(
        () =>
          ({
            getImageURI: jest.fn() as any,

            draw: draw as any,
          } as any)
      ),
      DataTable: jest.fn(
        () =>
          ({
            addColumn: jest.fn(),
            addRows: jest.fn(),
          } as any)
      ),
    },
    charts: {
      safeLoad: jest.fn(),
      load: jest.fn(),
      setOnLoadCallback: jest.fn((cb) => {
        cb();
      }),
    },
  };

  test("draw", () => {
    draw = jest.fn(() => {
      expect(draw.mock.calls[0]).toMatchSnapshot();
    });
    document.body.innerHTML = "<div id='chart'></div>";

    new Chart().init(document.body, [
      {
        lat: 50.444885,
        lng: 30.536667,
        flags: ["been"],
        city: "Kyiv(Kiev)",
        country: "Ukraine",
      },
      {
        lat: -34.45451,
        lng: -57.83755,
        flags: ["been"],
        city: "Colonia del Sacramento",
        country: "Uruguay",
      },
    ]);
  });

  test("getImage", async () => {
    draw = jest.fn(() => {});
    global.fetch = jest.fn().mockImplementation(() => ({
      blob: jest.fn(() => {
        return "blob";
      }),
    }));
    document.body.innerHTML = "<div id='chart'></div>";

    const res = await new Chart().init(document.body, [
      {
        lat: 50.444885,
        lng: 30.536667,
        flags: ["been"],
        city: "Kyiv(Kiev)",
        country: "Ukraine",
      },
      {
        lat: -34.45451,
        lng: -57.83755,
        flags: ["want"],
        city: "Colonia del Sacramento",
        country: "Uruguay",
      },
    ]);
    const blob = await res.getImage();
    expect(blob).toMatchInlineSnapshot(`"blob"`);
  });
});
