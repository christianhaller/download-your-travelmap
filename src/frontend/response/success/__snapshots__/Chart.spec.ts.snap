// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart draw 1`] = `
[
  {
    "addColumn": [MockFunction] {
      "calls": [
        [
          "number",
          "Lat",
        ],
        [
          "number",
          "Long",
        ],
        [
          "number",
          "Value",
        ],
        [
          {
            "role": "tooltip",
            "type": "string",
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": undefined,
        },
        {
          "type": "return",
          "value": undefined,
        },
        {
          "type": "return",
          "value": undefined,
        },
        {
          "type": "return",
          "value": undefined,
        },
      ],
    },
    "addRows": [MockFunction] {
      "calls": [
        [
          [
            [
              50.444885,
              30.536667,
              1,
              "Kyiv(Kiev)",
            ],
          ],
        ],
        [
          [
            [
              -34.45451,
              -57.83755,
              1,
              "Colonia del Sacramento",
            ],
          ],
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": undefined,
        },
        {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  {
    "backgroundColor": {
      "fill": "transparent",
      "stroke": "#FFF",
      "strokeWidth": 0,
    },
    "colorAxis": {
      "colors": [
        "#ff0099",
        "#000",
      ],
      "maxValue": 10,
      "minValue": 1,
    },
    "datalessRegionColor": "#f5f5f5",
    "displayMode": "markers",
    "keepAspectRatio": true,
    "legend": "none",
    "resolution": "countries",
    "sizeAxis": {
      "maxValue": 200,
      "minValue": 1,
    },
    "width": -40,
  },
]
`;
